### PR TITLE
Cuboid annotation

### PR DESCRIPTION
**Pull request by HollowTube**
Addressing https://github.com/opencv/cvat/issues/147

# Cuboid Annotation:

### Description
This PR adds fully functional cuboid annotation within CVAT.  The cuboid are fully integrated within CVAT and support regular features from other shapes such as copy-pasting, labels, etc. 

### Usage

Cuboid are created just like bounding boxes, simply select the cuboid shape in the UI and create. 
The cuboids may be edited by dragging certain edges, points or faces. Editing is constrained by a two point perspective model, that is, non-vertical edges all converge on either one of two vanishing points.

You may see these vanishing points in action by checking cuboid projection lines checkbox in the bottom left of the player.

### Annotation dump 
Points in the dump are ordered by vertical edges, starting with the leftmost edge and moving in counter clockwise order. The first point of each edge is always the top one. 

For example, the first point would be the top point of the leftmost edge and the second point would be the bottom point of the leftmost edge and the third point would be the top point of the edge in the front of the cuboid. 

### Known issues

* Currently, this build only supports dumping and uploading in cvat-xml format.
* The copy-paste buffer cuboid is just a polyline, but is still usable

The cuboids have been developed with the feedback of an in-house annotation team. This feature is fully functional but of course any feedback and or comment is appreciated!

### CODE REVIEW

**Comment by nmanovic** [comment]
@HollowTube , thanks for the PR! Great to see as a contributor Algolux company.

**Comment by nmanovic** [comment]
@HollowTube , do we really need math.js library? Do you use only "intersect" method to find a vanishing point? I'm fine with any 3rd-party library if it has a good license, popular, and really necessary. But if we can remove a 3rd-party dependency (especially if its size is more than 1M) it will improve loading time of CVAT slightly. What do you think?

**Comment by bsekachev** [comment]
@HollowTube @nmanovic 
Hi, first of all, thanks for you contribution. We need more time to review this PR carefully. But I have some first comments about it. 

- This patch is incompatible with Deep Extrem Cut plugin (It is DEXTR issue, but it also should be resolved from our side before merge)
- I am not sure how should I annotate with these Cuboids (maybe it's a matter of habit)
- Too many degrees of freedom. It's possible, to do different shapes (not only cuboids). Maybe it's good as a more common case, but it a little confuses. 
- It seems to be, that some corner cases during resize/drag aren't handled. Sometimes I can get strange results (attached below).
- There is minified math.js which has size about 138 KB (instead of current 1.8 MB). But it seems we don't want to have math.js as our source code (it has Apache 2.0 License). 
- We need to specify somewhere an order of points in a dumped file. Now 8 points are dumped, but we don't know what exactly is corner order.
Screenshot from 2019-08-30 13-51-37
Screenshot from 2019-08-30 13-52-13

**Comment by HollowTube** [comment]
@nmanovic 
Currently we are only using it for intersects, but I definitely could implement my own line intersection function. 

@bsekachev 

* Here's an example frame for cuboid annotation
image


* I could restrict editing in the UI to get only cuboids, but I was afraid it might negatively impact the UX. I'll toy around with restricting a few degrees of freedom

* I've specified the order of the points in the description, but im open to any suggestions on the dump format

* I am aware of the many edge cases, but in regular use they do not seem to be a problem. The annotators may come across these edge cases but they usually just delete the shape or undo as it does not completely crash the program. Could you elaborate on how you reached these edge cases?

Ill do a bit of refactoring along with some minor bug fixes in the next commit

**Comment by nmanovic** [comment]
@HollowTube , even Apache License is OK for us I would recommend to reimplement the intersect method (even to use existing code is a good practice big dependencies can lead to problems if we don't have strong reasons to integrate them).

I'm going to test the patch as well. It is a really important feature. Great to see the contribution!

Could you please also insert MIT header into new files?

```
/*
* SPDX-License-Identifier: MIT
*/
```

**Comment by nmanovic** [comment]
@HollowTube , my comments below. Let's discuss them one by one. If you like we can organize a call (I believe there is NDA between AlgoLux and Intel). I will be in the office on 9th September but Boris can definitely start the discussion without me. 

- Really easy to "break" cuboid. One way to break is to force a vanishing point to change its direction (sign). Even it doesn't crash the tool it will confuse users.
- If our frame has a non-zero roll angle it will be impossible to annotate correctly. At the moment vertical lines are parallel to the image boundaries always. Do you think it is a necessary restriction?
- It will be good to support interpolation of cuboids between frames. Let me know if you think you can do that. The cuboid has fixed number of points and really easy to translate from one key frame to another.
- I would imagine the cuboid as two quadrangles: front and back which are connected by edges. Thus we have 16 floats: (xtl1, ytl1, xtr1, ytr1, xbr1, ybr1, xbl1, ybl1, xtl2, ytl2, xtr2, ytr2, xbr2, ybr2, xbl2, ybl2). First is front and second is back. Thus it will be easy to interpolate each coordinate. Of course a user can swap first and second quadrangles or swap top and bottom edges during drawing but order of coordinates should be adjusted respectively (e.g. it isn't important how we draw a box. We always correctly define xtl, ytl, xtr, ytr, xbr, ybr, xbl, ybl).
- Algorithm which calculates invisible edges works incorrectly in some cases.

Questions:
- What is the current annotation speed for cuboids?
- Do you use it only for ADAS to annotate vehicles? What is about robotic use case? Will the proposed approach flexible enough to other use cases?]
- How many cuboids did you already annotate?
- Do you need to annotate only images (e.g. for detection)? Did you try to annotate video files (e.g. for tracking)?
- Do you know how other tools implement the feature? Any examples? Did you base on the experience?

**Comment by timbowl** [comment]
@nmanovic , sorry for the delay in replying. @HollowTube is away this week, but I can address a few of your comments. (I'm his colleague at Algolux)

- Really easy to "break" cuboid: Before he left, he had a few ideas on how to restrict editing to ensure valid cuboids.
- Non-zero roll angle: For us, this was corner case we didn't implement in order to simplify the implementation. I do agree it can be useful, though. Can this be something implemented afterwards?
- Cuboid interpolation: Yes, we'd love to have this feature eventually, but it wasn't a requirement for us initially. We'll take a closer look at your implementation proposal. 

When @HollowTube returns next week, we'll be happy to set up a conference call. We can also bring in our annotation team manager and he can answer you questions regarding how our annotators are currently using the cuboid feature.

We're really looking forward to working with you!

**Comment by bsekachev** [comment]
A little more comments here:
1) Occluded points shouldn't be dashed (the screenshot below)
2) I have tried to merge a couple of cuboids and the tool broke (I can't do something after)
3) Cuboids are changeable even if they are locked (Lock doesn't work)

Code looks well. 

Screenshot from 2019-09-10 15-30-05

File: cvat/apps/engine/static/engine/js/annotationParser.js [hunk 1]
```diff
@@ -155,6 +155,7 @@ class AnnotationParser {
             polygons: [],
```
[the remark this answers is not in the document]
Comment (HollowTube) [reply cvat/apps/engine/static/engine/js/annotationParser.js]: Yes I have noticed, I changed cvat.py appropriately. Will also be adding the new naming convention for points soon.

**Comment by timbowl** [comment]
> @timbowl , @HollowTube ,
> 
> > When @HollowTube returns next week, we'll be happy to set up a conference call. We can also bring in our annotation team manager and he can answer you questions regarding how our annotators are currently using the cuboid feature.
> 
> Let me know when you are available. I will be glad to discuss the feature on a conference call. What do you think? Please contact me "[first name].[last name] at intel dot com"

Sounds good. Once I get everyone's availability, we'll get back to you.

**Comment by timbowl** [comment]
Someone on the public Gitter forum mentioned the need for shapes that extend beyond the frame boundary. We had to modify our fork to support this use case only for cuboids, but we had to remove the `min_value=0` restriction from `ShapeSerializer`. This is not ideal as it affects all shapes. Not sure what the best way to go about this... (note that this change is not in this PR)

**Comment by HollowTube** [comment]
@nmanovic @bsekachev 

I've fixed some of the issues you've mentioned.
* The occlusion should work properly now
* The MIT License has been adding to files that were missing it
* Cuboids are locked properly now
* VsCode settings file has been reverted
* Points in the dump file are now named using your suggestion. 

I've also adding some constraints to the resizing of the left and right vertical edges. You should not be able to to switch the direction of the the vanishing through them. However it is still possible to get non-cuboids by resizing the middle edge and dragging the back faces.   

The reason non-cuboids shapes happen is that the vanishing point are not lined up on the same horizon. It is of course possible to enforce this but this would break one-point perspective cuboids which make up a large portion of our use cases. What would be an acceptable degree of freedom? 

Also, as @timbowl mentioned, our use cases requires us to draw cuboids that are partially outside of the frame. What are the dangers of doing this?

Still working on fixing the merging for cuboids

File: .vscode/settings.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-    "python.pythonPath": ".env/bin/python",
+    "python.pythonPath": "/home/tritin.truong/.virtualenvs/eat-2.0/bin/python",
```
[the remark this answers is not in the document]
Comment (jriddy) [reply .vscode/settings.json]: Shouldn't IDE settings files like `.vscode` be removed entirely from the repo and excluded in the future via `.gitignore`? I believe that's the usual way of handling things like this since it's gonna depend on a each developers local config.

[the remark this answers is not in the document]
Comment (bsekachev) [reply .vscode/settings.json]: @jriddy

No, these files shouldn't be removed because [this manual](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md) about how to setup dev environment is based on it. 
So, if we change/delete this settings, the steps from the manual won't work as expected.

[the remark this answers is not in the document]
Comment (timbowl) [reply .vscode/settings.json]: I also found it useful to recreate the launch scripts on PyCharm.

**Comment by nmanovic** [comment]
@HollowTube ,

Let me mention a couple of usability issues which we discussed on our meeting:
- Cursor view should correspond to an action which users want to do (hint). Look at how regular bounding boxes do that. For example, when we try to adjust an edge it has to show possible options (up and down arrow, left and right arrow). Now it shows move cursor always. 
- For bounding boxes we have explicit points which you can grab and adjust an edge. I believe the same concept will work for 3D bounding boxes. Just add a corresponding point in the middle of some edges. Look at how 2D bounding boxes work.
image

- Need to fix appearance of visual edges. They should not be dashed.
image

- Sometimes I cannot move an edge of 3D bounding box out of the image, sometimes I can do. Could you please explain or fix the logic?

- Copy/paste visualization for 3D bounding boxes can be improved
image

- May be orientation by default isn't a good choice. Another way is to use orientation of the previous 3D box.

- Can we move "Cuboid Projection Lines" to settings (F2)? I don't think that it should be changed often. 

- It is easy to create a 3D shape like on the screenshot below. Probably if we have an option in context menu (right button click) to normalize (make a parallelepiped) the shape it will be awesome. It is 
image

**Comment by nmanovic** [comment]
@HollowTube , one more idea about changing geometry of the 3D bounding box. We always can divide the 3D box on two trapezoids (near and far). Right? It doesn't matter they are (side, side) or (front, back). In the example below (side, side).
image

I like that I can drag bottom and top edges of the near trapezoid and change height. But for left and right edges of the near trapezoid I cannot change width. For the example above I can change width only using right edge of the near trapezoid.

The main point of the experiment above is to motivate you to describe different actions which you can do with visible/invisible edges, corners, planes as a reply on the comment. Thus you will easily find some inconsistency and can improve UI/UX.

**Comment by nmanovic** [comment]
@bsekachev , do you have any comments about the code?

**Comment by bsekachev** [comment]
@nmanovic 
No, I don't have other comments about the code.

**Comment by HollowTube** [comment]
@nmanovic 
 
Refering to the UX of the cuboid: 

The middle vertical edge is very tricky and I've resisted adding drag options for a few reasons. How would you know if the vertical edge should adjust the left or right face? In your example, it seems very intuitive that the middle edge should adjust the right box, but what if the orientation of the car were different? The side face would be to the left and so dragging the middle edge should adjust the width of the left face. 

Right now the behaviour of the middle edge is consistent although a little un-intuitive. However, I am unsure on how to implement a consistent dragging behaviour. Should it depend on the current dimensions of the cuboid? Should both options be available at all times on the UI? If so how should this be communicated to the user? 

A possible solution could be to give a default dragging direction and modify this direction by holding down some key like "~". I think this is the most efficient option for advanced users, but it may take getting used to for novice users. Let me know your thoughts on this. 

Ill should be able to fix most of the minor issues you have except a few:
* Dynamically changing the default orientation of the cuboid is difficult as it would introduce a dependency between the shape creator and the actual shape instantiated 
* Normalizing the shape of the cuboid to a parallelepiped is possible, but would take more time, Ill try my best to get it done but it may be pushed to another PR.

**Comment by nmanovic** [comment]
@HollowTube ,

Could you please fix the issue with build?

```
CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0022_auto_20191004_0817, 0022_auto_20190829_1911 in engine).
To fix them run 'python manage.py makemigrations --merge'
The command "docker exec -it cvat /bin/bash -c 'python3 manage.py test cvat/apps/engine utils/cli'" exited with 1.
```

**Comment by jriddy** [comment]
@nmanovic I invited you as a collaborator on our private repo, let me know if there are any additional permissions issues.

**Comment by nmanovic** [comment]
@jriddy, @HollowTube ,

For you convenient I summarized all ideas which were discussed previously. Let's implement them and after that merge the PR if we don't see obvious and trivial ways to enhance it.

*UI tricks*
- Don’t use vanishing points to draw a cuboid.
- Cursor view should correspond to an action which users want to do (hint). Look at how regular bounding boxes do that. For example, when we try to adjust an edge it has to show possible options (up and down arrow, left and right arrow). Now it shows move cursor mostly always.
- For bounding boxes we have explicit points which you can grab and adjust an edge. I believe the same concept will work for 3D bounding boxes. Just add a corresponding point in the middle of some edges. Look at how 2D bounding boxes work.

*Bugs*
- Merge doesn’t work as Boris said
c56de67398f0.js:6881 Uncaught Error: Unreacheable code was reached.
    at buildShapeModel (c56de67398f0.js:6881)
    at ShapeCollectionModel.add (c56de67398f0.js:8432)
    at ShapeMergerModel.done (c56de67398f0.js:10837)
    at ShapeMergerController.switch (c56de67398f0.js:10922)
    at HTMLButtonElement.<anonymous> (c56de67398f0.js:10939)
    at HTMLButtonElement.dispatch (89deac3c33bb.js:23468)
    at HTMLButtonElement.elemData.handle (89deac3c33bb.js:23276)

*ZBB*
- interpolation
- non-zero roll

**Comment by HollowTube** [comment]
@nmanovic 
Thank you for the suggestions, Im currently working on a simpler model with only a single vanishing point and reworking the way the cuboid is edited. This model will hopefully be simpler to annotate at the cost of some flexibility. 
Would it be acceptable to have 2 modes of editing down the line? Something like a simple mode that is fast and easy to annotate, but less flexible, and a complex and slow mode which would allow you to have much greater control over the cuboid at the cost of speed.

I have been very busy lately, but I hope to be done within 2 weeks with this experiment as well as some of the fixes you mentioned

**Comment by nmanovic** [comment]
> @nmanovic
> Thank you for the suggestions, Im currently working on a simpler model with only a single vanishing point and reworking the way the cuboid is edited. This model will hopefully be simpler to annotate at the cost of some flexibility.
> Would it be acceptable to have 2 modes of editing down the line? Something like a simple mode that is fast and easy to annotate, but less flexible, and a complex and slow mode which would allow you to have much greater control over the cuboid at the cost of speed.
> 
> I have been very busy lately, but I hope to be done within 2 weeks with this experiment as well as some of the fixes you mentioned

At the moment I would recommend to concentrate on one mode only. It will be simpler.

**Comment by nmanovic** [comment]
Hi @HollowTube ,

Any progress with the pull request? Soon we are going to merge a big patch. It will be a huge challenge to merge the PR after that. I think that at the end of the year it will be mostly impossible to merge the PR. If you want to contribute and offload the feature support on our side it will be great to merge the patch till end of November. Please let me know your decision.

**Comment by HollowTube** [comment]
@nmanovic 
Hi, sorry for the delay, I have been able to fix the UI cursors along with some grab points, but I am still developing the model with less vanishing points. I still want to merge it. Would it be acceptable to merge it with just the UI fixes and update the vanishing point model later on?

**Comment by HollowTube** [comment]
@nmanovic 
This is the version I have been working on, it still uses vanishing points to draw the cuboid, but the control scheme has been improved. Every edge can now be dragged and every face except the front face can be dragged. Try out these new controls and let me know what you think. I have also included the cursor hints and the UI grab points. Let me know what you think.

**Comment by nmanovic** [comment]
@HollowTube , I have found an issue. When I tried to download annotations in "CVAT for videos" format and upload them again I got 500 error:

<details>

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/rq/worker.py", line 812, in perform_job
    rv = job.perform()
  File "/usr/local/lib/python3.5/dist-packages/rq/job.py", line 588, in perform
    self._result = self._execute()
  File "/usr/local/lib/python3.5/dist-packages/rq/job.py", line 594, in _execute
    return self.func(*self.args, **self.kwargs)
  File "/usr/lib/python3.5/contextlib.py", line 30, in inner
    return func(*args, **kwds)
  File "/home/django/cvat/apps/engine/annotation.py", line 112, in load_task_data
    annotation.upload(filename, loader)
  File "/home/django/cvat/apps/engine/annotation.py", line 760, in upload
    execute_python_code("{}(file_object, annotations)".format(loader.handler), global_vars)
  File "/home/django/cvat/apps/engine/utils.py", line 60, in execute_python_code
    raise InterpreterError("{} at line {}: {}".format(error_class, line_number, details))
cvat.apps.engine.utils.InterpreterError: KeyError at line 435: xtl1
```

</details>

**Comment by nmanovic** [comment]
Can we draw points in the case on the visiable side? Otherwise it is more convent to annotate vehicles only if they are on left side.

image

**Comment by nmanovic** [comment]
It is a minor comment but can we draw only boxes for back and side? I mean if we edit a 3d box or it is active let's show all edges but I believe it will be more natural to draw only visible sides in the case. What do you think?

image

**Comment by nmanovic** [comment]
@HollowTube , don't need to do anything special to break a cuboid. I just pull down a corner.

image

**Comment by nmanovic** [comment]
@HollowTube , I got the crash in console. I don't know how to reproduce:

```
daceb0c4bab5.js:17449 Uncaught TypeError: Cannot read property 'x' of null
    at PaintHandler.calc (daceb0c4bab5.js:17449)
    at PaintHandler.done (daceb0c4bab5.js:17325)
    at initializer.draw (daceb0c4bab5.js:17402)
    at ShapeCreatorView.onShapeCreatorUpdate (20a76aa8528b.js:11643)
    at ShapeCreatorModel.notify (20a76aa8528b.js:1278)
    at ShapeCreatorModel.switchCreateMode (20a76aa8528b.js:11138)
    at ShapeCreatorController.switchCreateMode (20a76aa8528b.js:11190)
    at ShapeCreatorController.<anonymous> (20a76aa8528b.js:11182)
    at ShapeCreatorController.<anonymous> (20a76aa8528b.js:1111)
    at _fireCallback (daceb0c4bab5.js:11147)
```

**Comment by HollowTube** [comment]
@nmanovic 
Hi, its possible to put the points on the left side, but that may take some more time, I can fix all the critical issues soon. Could you outline in more detail how you are breaking the cuboid, I am having trouble reproducing it on my side. Same for the browser crashing, just a simple context would help a lot.

**Comment by nmanovic** [comment]
@HollowTube , please merge the latest develop branch.

Please see below how to reproduce the broken cuboid:

broken_cuboid

**Comment by nmanovic** [comment]
If you draw a cuboid the statistics is incorrect (2 instead of 1 cuboid):
image

**Comment by nmanovic** [comment]
I don't understand the algorithm how highlighting works.
highlight_side

**Comment by nmanovic** [comment]
I'm trying to reproduce crashes on the server and client side.

**Comment by nmanovic** [comment]
@HollowTube , 

I really like the feature and you willing to help us but I don't see that the patch is going to be finished in the nearest future. I will propose to use the PR as a future guidance on how the future can be implemented.

In general I have many concerns about the patch:
- I don't like how the cuboid is drawing. From one side it doesn't cover the common case but at the same time it is not convenient to annotate vehicles on the road as well. I have a couple of ideas how to draw them but to prove it is necessary to implement these options and try. Look at the gif below. First part should be drawn by a human, second part can be automatically drawn by the algorithm. After that need slightly update the cuboid. The idea below can cover the common case and it seems it is more convenient for the road specific case as well (but it is only an idea which comes first after you annotate enough cuboids).

cuboid

- If I touch one point it can affect location of others (it is hard to predict and control)
- It doesn't implement interpolation (probably can wait)
- Mouse hints are good but probably can be improved slightly
- You can easily draw a cuboid in the patch and it is a problem which annotators will hate.
- Vanish points should be eliminated. It is evil which just make things worse.

**Comment by HollowTube** [comment]
@nmanovic 

Thank you for your sincere feedback. I remain commited to contributing to cvat and will soon have much more time to work on it full time over the winter break to address the changes you have proposed.

Here are some of the changes that I can implement:
* Get rid of the vanishing point model entirely, this will also stop the adjustment of other points while editing
* Fix the breaking cuboid issue
* Fix the statistics bug  
* Change the cuboid drawing

In the very early stages, I drew the cuboid in a manner similar to what you proposed, I can try going back to that method of drawing if you wish. I changed it to the current iteration because I found it far too time consuming to create a single cuboid, so i switched it to the current version as a trade-off between the time adjusting the cuboid during creation and during editing.  

If this is something you are still interested in, please let me know.

**Comment by nmanovic** [comment]
> In the very early stages, I drew the cuboid in a manner similar to what you proposed, I can try going back to that method of drawing if you wish. I changed it to the current iteration because I found it far too time consuming to create a single cuboid, so i switched it to the current version as a trade-off between the time adjusting the cuboid during creation and during editing.
> 
> If this is something you are still interested in, please let me know.

I'm going to wait as much as necessary. We will help you to adopt your patch in the future. Let's keep the PR open till you have time to contribute or our team has resources to integrate the great contribution!

**Comment by HollowTube** [comment]
@nmanovic 

Hello, I have integrated the changes you have suggested for this iteration. 
For starters, I have changed the drawing method to what you have suggested.

Here it is in action.
with_perspective

From another another angle:
with_perspective2

As you can see, it is able to handle building the cuboid as you specified for both the left and the right orientations. The editable back edge also adjusts dynamically depending on where it is situated. For now, you must draw the points in the order as shown in the demo above for the most consistent results, that is, by drawing rotated 'L' shape. Let me know how it feels.  

The statistics bug should also be fixed.

There are still some perspective effects included (i.e.  vanishing points), but I think they are necessary for even the most basic of use cases, as seen below. 

Without any perspective effects you get this:
Screenshot from 2019-12-15 22-15-05
However with perspectives effects you are able to get this:
Screenshot from 2019-12-15 22-16-25

As you can see, without any perspective effects, the cuboid does not fit the car very well compared to the one with the perspective effects. I effectively use a single vanishing point now so it should be much more stable and harder to break.

If you still want to get rid of the vanishing points, I can do it easily.
Let me know what you think about this iteration, and if there is anything I can do to further improve it.

**Comment by nmanovic** [comment]
@HollowTube ,

First of all it is a great job! Really huge progress in the user experience. It feels very well. It isn't just words. I really think so!

I have an idea how to improve the user experience further. First all `Exactly 4 points must be used for cuboid drawing and must be drawn in a rotated 'L' shape` is a very confusing message.

* if it is necessary only 4 points why don't try to finish drawing exactly after the user puts 4 points?
* Also I can suggest another analogy. We need to put 3 points on a visible plane to identify it and draw a perpendicular to the plane. Does it sound more natural?
* Now it is mandatory to draw in some "pre-defined" order. Otherwise I get a wrong cuboid. With the explanation above it is easy to draw in any order and you should be able to complete the cuboid and understand its orientation. Also in this case you can draw "arbitrary" cuboid with sides which don't parallel to the image borders.
* It will be good to have a special option. Keep a key on the keyboard (e.g. shift) and put two points to draw a bounding box (identify a plane which parallels to the camera plane). After that draw a "perpendicular" to the image plane. Thus we will be able to draw a cuboid using only 3 points.
*  Vanishing point mode is necessary. I agree. Probably it can be activated by a shortcut (e.g. keep shift).

**Comment by HollowTube** [comment]
@nmanovic 

Thank you for your kind words!

* I can definitely try to make the it stop exactly after 4 points. 
* I'm not sure I understand your analogy, right now it is effectively always using the first 3 points to define the front face and the last point is used to determine the depth and the orientation or the cuboid. Do you mean to choose any 3 of 4 points to define the face and choose another point to determine the depth? In that case, where would the base for the perpendicular be, the middle of the plane, one of the defined points? 
* I like the idea of the special option and I would definitely like to implement it in the future, but right now it would require me to create a custom drawing method, so I hope it can be implemented in a future PR.  
* How would you limit the vanishing point setting? Do you mean hold shift during creation time to create the cuboid with vanishing points? Or do you mean only reveal the option to reveal the editable back point during editing time by holding shift? I dont think it is necessary in either case as I believe users should be able to grasp this simple 3D perspective concept quickly. 

Here are is my plan:
* I will force the cuboid drawing to stop at 4 points
* I will change the cuboid error message to something more reasonable
* I will fix some formatting and code style to pass the CI tests. 

What else would you require before a merge?

**Comment by nmanovic** [comment]
@HollowTube ,

As I said I'm already feeling the the user experience is awesome. Really great and in the update it is even better.

> I can definitely try to make the it stop exactly after 4 points.

Well done! It works.

good_cuboid1

> I'm not sure I understand your analogy, right now it is effectively always using the first 3 points to define the front face and the last point is used to determine the depth and the orientation or the cuboid. Do you mean to choose any 3 of 4 points to define the face and choose another point to determine the depth? In that case, where would the base for the perpendicular be, the middle of the plane, one of the defined points?

I mean to choose any 3 of 4 points to define a plane and choose another point to determine the depth. 

About "face plain". How to choose it? How to specify front of a car? Let's discuss in a future. It isn't important for the PR.

Look at some problems with the current approach. If I try to draw a 3D box in a way you don't expect I get strange results. If you take "any 3 of 4 points" approach it should work much better.

wrong_cuboid1

wrong_cuboid2

> I like the idea of the special option and I would definitely like to implement it in the future, but right now it would require me to create a custom drawing method, so I hope it can be implemented in a future PR.

OK.

> How would you limit the vanishing point setting? Do you mean hold shift during creation time to create the cuboid with vanishing points? Or do you mean only reveal the option to reveal the editable back point during editing time by holding shift? I dont think it is necessary in either case as I believe users should be able to grasp this simple 3D perspective concept quickly.

I mean when you are moving a point (look below) try to produce a perspective effect only if shift is pressed. 


On the gif below you can see some artefacts: jittering. Also it is better to use double arrows when it is applicable because you can move a point in both directions.

cuboid_bug1

**Comment by nmanovic** [comment]
Please fix codacy issues. It checks possible problems and verify our coding guidelines: https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=4086021

**Comment by nmanovic** [comment]
@HollowTube ,

Happy New Year! Do you think it is visible to solve mentioned issues soon?

**Comment by HollowTube** [comment]
@nmanovic 
Yes! It should be done by tomorrow. I have also already included a feature to allow users to use the perspective mode only when they hold shift.
What else would you require?

**Comment by HollowTube** [comment]
@nmanovic 
The codacy issues have been fixed, not sure why the travis build is failing

**Comment by nmanovic** [comment]
@HollowTube ,

it is failing because of a conflict in migrations. Need to fix.
```
Creating test database for alias 'default'...
5571CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0023_auto_20200113_1323, 0023_auto_20191028_1138 in engine).
```

**Comment by nmanovic** [comment]
@HollowTube , 3 issues which I described in my early comments are still here:

**Comment by HollowTube** [comment]
@nmanovic 
Hi, I've addressed the first 2 issues. I've fixed the jittering. You may now pick any 3 of the 4 points of the pink face and choose the last point for depth and orientation. I think this will satisfy most use cases for now, and I do not think it is necessary to implement drawing the cuboid from the top face as this is un-intuitive, what do you think?

**Comment by nmanovic** [comment]
> @nmanovic
> Hi, I've addressed the first 2 issues. I've fixed the jittering. You may now pick any 3 of the 4 points of the pink face and choose the last point for depth and orientation. I think this will satisfy most use cases for now, and I do not think it is necessary to implement drawing the cuboid from the top face as this is un-intuitive, what do you think?

@HollowTube , thanks for your willing to improve the PR. We are mostly in the final point. Could you please clarify why `top face` case is more difficult to support than the other cases? I don't think that it is un-intuitive. For me the logic is very trivial: choose any 3 point to define a plane and draw a perpendicular to define the depth. It is easy to explain and remember. More difficult to understand why it doesn't work in `top face` case.

**Comment by HollowTube** [comment]
@nmanovic 
I implement what you asked for, it does work, but it is a little strange for the user as in, they need to draw the last line almost exactly downwards or upwards. The difficulty was not in drawing the cuboid, but in organizing the points. My other worry was that perhaps the behaviour would be a little too inconsistent for the user.
Try it out and let me know what you think, so that we can get this merged as soon as we can!

**Comment by nmanovic** [comment]
@HollowTube , I can confirm that now it works as it is expected (at least by myself). Another problem, "shift" doesn't work in some cases. Look at two cases below. In one case "shift" works as expected, but if I draw the cuboid in a different direction it doesn't work. What is the right approach here from you point of view? Is it a bug?



shift_1

shift_2

**Comment by HollowTube** [comment]
@nmanovic 
This is an expected behaviour, because I am only using a single vanishing point that is located between the pink face and its opposite face. It's possible to make it possible to make it work for both, but this would require bringing back a 2 point perspective model. To make it work, the pink face should always be the closest face. When drawing the first plane they draw will always be the pink face, unless it is the top or bottom face, in which case it will be the left face by default. 
In my opinion, this pink face should be enough for users to know that this face is special and follows other rules. I don't think this should be a problem to explain with enough instruction or training.

**Comment by nmanovic** [comment]
@HollowTube , Why do we need the limitation? Is it more obvious to turn on "vanishing point" mode using "shift"?

**Comment by HollowTube** [comment]
@nmanovic 
Vanishing point is not a mode per say, the vanishing point is always there, its just that when you hold shift, the controls allow you to change the length of that edge while following the constraints of the vanishing point.
I designed the editing of the cuboid specifically around that one vanishing point for simplicity. If you wanted to change every face with perspective, then I would have to redesign the editing again around 2 vanishing points or even more, and that is what led to an easily broken cuboid.

**Comment by nmanovic** [comment]
@HollowTube ,  if you look on all my comments you will see that they are about simplification and predictability. Now it is obvious how to draw a cuboid from my point of view (or easy to explain and remember). I like your "shift" implementation but I don't understand why it depends on the order of points. In general after I draw a cuboid I want to adjust its perspective in one or two directions and it depends on the object orientation.

I think that the main problem with your previous implementation (when you used 2 vanishing points) was with roll == 0 hypothesis. Thus with only one vanishing point you defined the horizontal line on the image (it had to be parallel to x-coordinate).

I see two solutions here:
1. It is easy to implement and probably will close most of UX problems. Keep roll == 0 constraint as is. Allow to adjust perspective only in one direction (which the user will choose). Thus if you try to adjust for the same cuboid perspective in another direction we will notify that it is impossible. Also *need to have an option in the context menu for a specific cuboid to reset perspective effect*.
2. Remove roll == 0 constraint and you will be able to draw two vanishing points and they will define the image horizontal line (which isn't parallel to x-coordinate in common case). Also it will change direction of your cuboid sides (which are not parallel to y-coordinate in common case). We even can have settings per image to adjust roll (by default it can be zero). It is the most general solution but it requires more work and experiments.

**Comment by HollowTube** [comment]
@nmanovic 
Understood, I think I will implement option 1 for now, how do you envision choosing the perspective face for the user? Would a simple shortcut to swap the perspective face be enough as well as in the context menu? Is the default perspective face good enough for now? I will also be implementing the reset for specific cuboids.

**Comment by nmanovic** [comment]
> @nmanovic
> Understood, I think I will implement option 1 for now, how do you envision choosing the perspective face for the user? Would a simple shortcut to swap the perspective face be enough as well as in the context menu? Is the default perspective face good enough for now? I will also be implementing the reset for specific cuboids.

@HollowTube , I expect:
- User can adjust perspective in one direction. It should work with "shift" button as now. It should not depends on order of points.

shift_2

- User should be able to reset the perspective effect. Click on the right mouse button. Choose "Reset perspective".

- If a user tries to change perspective in a second direction "shift" button doesn't have any effect. Need to reset perspective.

- A cuboid highlights a side. The highlighted side is saved as the first bounding box. I will propose don't allow a user to change it for now. Let's implement the feature later when somebody explicitly requests it.

**Comment by HollowTube** [comment]
@nmanovic 
The changes should be done, the user can both reset the perspective and choose which face perspective face to use from the context menu.
Fixing the code for codacy soon. Do you know why it is suddenly highlighting errors in files i have never touched? Which eslint file codacy is using? I cant seem to use the correct file locally.

**Comment by VermaAnkit** [comment]
Hi, nikita sir how to add cubiod feature on cvat help  me

**Comment by nmanovic** [comment]
@HollowTube , we have some codacy issues. Don't spend your time to fix codacy issues for now. Please fix merge conflicts.

**Comment by HollowTube** [comment]
@nmanovic 
The merge is complete, but the codacy is failing, as expected. Do tell if there are any breaking bugs still.

**Comment by nmanovic** [comment]
@azhavoro , could you please try the PR and share your opinion?

**Comment by nmanovic** [comment]
@azhavoro , any comments? Does it work for you?

**Comment by azhavoro** [comment]
> @azhavoro , any comments? Does it work for you?

@nmanovic I've tested patch, generally it works well, but need documentation for this feature. I mean how to add and edit the cuboid shape at least. Do we plan to write it internally?

File: .vscode/settings.json [hunk 3]
```diff
@@ -1,5 +1,5 @@
 {
-    "python.pythonPath": ".env/bin/python",
+    "python.pythonPath": "/home/tritin/.venvs/cvat/bin/python",
```
Comment (azhavoro) [inline .vscode/settings.json]: Please revert this change.

**Comment by nmanovic** [comment]
> > @azhavoro , any comments? Does it work for you?
> 
> @nmanovic I've tested patch, generally it works well, but need documentation for this feature. I mean how to add and edit the cuboid shape at least. Do we plan to write it internally?

 Yes. We are going to write it internally.

**Comment by HollowTube** [comment]
@nmanovic 
The documentation should be written in user_guide.md yes? I will try to write up a simple one.

**Comment by nmanovic** [comment]
@HollowTube , let's merge it. Unfortunately it will take time to port the feature in the new UI. Thus it will be available in legacy UI for a while.
